File: dadakar-front-admin/src/main/java/com/umanteam/dadakar/admin/front/service/implementation/MessageService.java
package com.umanteam.dadakar.admin.front.service.implementation;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpMethod;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.util.MultiValueMap;
import org.springframework.web.client.RestTemplate;

import com.umanteam.dadakar.admin.front.DadakarFrontAdminApplication;
import com.umanteam.dadakar.admin.front.dto.Message;
import com.umanteam.dadakar.admin.front.service.interfaces.IMessageService;

@Service
public class MessageService implements IMessageService {
	
	@Autowired
	private RestTemplate restTemplate;
	
	@Value("${msg.path}")
	private String msgPath;

	@Override
	public Message add(Message message) {
		MultiValueMap<String, String> headers = new LinkedMultiValueMap<>();
		headers.add("Authorization", DadakarFrontAdminApplication.getToken());
		HttpEntity<Message> msgRequest = new HttpEntity<Message>(message, headers);
		ResponseEntity<Message> msgResponse = restTemplate.exchange(msgPath + "/save", HttpMethod.POST, msgRequest, Message.class);
		if(msgResponse.getStatusCode() != HttpStatus.OK) return null;
		message = msgResponse.getBody();
		return message;
	}
	
	@Override
	public Message update(Message message) {
		MultiValueMap<String, String> headers = new LinkedMultiValueMap<>();
		headers.add("Authorization", DadakarFrontAdminApplication.getToken());
		HttpEntity<Message> msgRequest = new HttpEntity<Message>(message, headers);
		ResponseEntity<Message> msgResponse = restTemplate.exchange(msgPath + "/update", HttpMethod.PUT, msgRequest, Message.class);
		if(msgResponse.getStatusCode() != HttpStatus.OK) return null;
		message = msgResponse.getBody();
		return message;
	}

	@Override
	public void delete(String id) {
		Message message = new Message();
		MultiValueMap<String, String> headers = new LinkedMultiValueMap<>();
		headers.add("Authorization", DadakarFrontAdminApplication.getToken());
		HttpEntity<Message> msgRequest = new HttpEntity<Message>(message, headers);
		restTemplate.exchange(msgPath + "/del/" + id, HttpMethod.DELETE, msgRequest, Message.class);
	}

	@Override
	public List<Message> findAll() {
		List<Message> messages = new ArrayList<>();
		MultiValueMap<String, String> headers = new LinkedMultiValueMap<>();
		headers.add("Authorization", DadakarFrontAdminApplication.getToken());
		HttpEntity<List<Message>> msgRequest = new HttpEntity<List<Message>>(messages, headers);
		ResponseEntity<List<Message>> msgResponse = restTemplate.exchange(msgPath, HttpMethod.GET, msgRequest, new ParameterizedTypeReference<List<Message>>() {});
		messages = msgResponse.getBody();
		return messages;
	}

	@Override
	public Message findById(String id) {
		String url = msgPath + "/" + id;
		Message message = new Message();
		MultiValueMap<String, String> headers = new LinkedMultiValueMap<>();
		headers.add("Authorization", DadakarFrontAdminApplication.getToken());
		HttpEntity<Message> msgRequest = new HttpEntity<Message>(message, headers);
		ResponseEntity<Message> msgResponse = restTemplate.exchange(url, HttpMethod.GET, msgRequest, Message.class);
		message = msgResponse.getBody();
		return message;
	}

	@Override
	public List<Message> findByHoroLessThanEqual(LocalDateTime horoEnd) {
		String url = msgPath + "/lte:" + horoEnd.toString();
		List<Message> messages = new ArrayList<>();
		MultiValueMap<String, String> headers = new LinkedMultiValueMap<>();
		headers.add("Authorization", DadakarFrontAdminApplication.getToken());
		HttpEntity<List<Message>> msgRequest = new HttpEntity<List<Message>>(messages, headers);
		ResponseEntity<List<Message>> msgResponse = restTemplate.exchange(url, HttpMethod.GET, msgRequest, new ParameterizedTypeReference<List<Message>>() {});
		messages = msgResponse.getBody();
		return messages;
	}

	@Override
	public List<Message> findByHoroGreaterThanEqual(LocalDateTime horoStart) {
		String url = msgPath + "/gte:" + horoStart.toString();
		List<Message> messages = new ArrayList<>();
		MultiValueMap<String, String> headers = new LinkedMultiValueMap<>();
		headers.add("Authorization", DadakarFrontAdminApplication.getToken());
		HttpEntity<List<Message>> msgRequest = new HttpEntity<List<Message>>(messages, headers);
		ResponseEntity<List<Message>> msgResponse = restTemplate.exchange(url, HttpMethod.GET, msgRequest, new ParameterizedTypeReference<List<Message>>() {});
		messages = msgResponse.getBody();
		return messages;
	}

	@Override
	public List<Message> findByHoroBetween(LocalDateTime horoStart, LocalDateTime horoEnd) {
		String url = msgPath + "/gte:" + horoStart.toString() + "/lte:" + horoEnd.toString();
		List<Message> messages = new ArrayList<>();
		MultiValueMap<String, String> headers = new LinkedMultiValueMap<>();
		headers.add("Authorization", DadakarFrontAdminApplication.getToken());
		HttpEntity<List<Message>> msgRequest = new HttpEntity<List<Message>>(messages, headers);
		ResponseEntity<List<Message>> msgResponse = restTemplate.exchange(url, HttpMethod.GET, msgRequest, new ParameterizedTypeReference<List<Message>>() {});
		messages = msgResponse.getBody();
		return messages;
	}

	@Override
	public List<Message> findBySenderId(String sid) {
		String url = msgPath + "/sid:" + sid;
		List<Message> messages = new ArrayList<>();
		MultiValueMap<String, String> headers = new LinkedMultiValueMap<>();
		headers.add("Authorization", DadakarFrontAdminApplication.getToken());
		HttpEntity<List<Message>> msgRequest = new HttpEntity<List<Message>>(messages, headers);
		ResponseEntity<List<Message>> msgResponse = restTemplate.exchange(url, HttpMethod.GET, msgRequest, new ParameterizedTypeReference<List<Message>>() {});
		messages = msgResponse.getBody();
		return messages;
	}

	@Override
	public List<Message> findBySenderIdAndHoroLessThanEqual(String sid, LocalDateTime horoEnd) {
		String url = msgPath + "/sid:" + sid + "/lte:" + horoEnd.toString();
		List<Message> messages = new ArrayList<>();
		MultiValueMap<String, String> headers = new LinkedMultiValueMap<>();
		headers.add("Authorization", DadakarFrontAdminApplication.getToken());
		HttpEntity<List<Message>> msgRequest = new HttpEntity<List<Message>>(messages, headers);
		ResponseEntity<List<Message>> msgResponse = restTemplate.exchange(url, HttpMethod.GET, msgRequest, new ParameterizedTypeReference<List<Message>>() {});
		messages = msgResponse.getBody();
		return messages;
	}

	@Override
	public List<Message> findBySenderIdAndHoroGreaterThanEqual(String sid, LocalDateTime horoStart) {
		String url = msgPath + "/sid:" + sid + "/lte:" + horoStart.toString();
		List<Message> messages = new ArrayList<>();
		MultiValueMap<String, String> headers = new LinkedMultiValueMap<>();
		headers.add("Authorization", DadakarFrontAdminApplication.getToken());
		HttpEntity<List<Message>> msgRequest = new HttpEntity<List<Message>>(messages, headers);
		ResponseEntity<List<Message>> msgResponse = restTemplate.exchange(url, HttpMethod.GET, msgRequest, new ParameterizedTypeReference<List<Message>>() {});
		messages = msgResponse.getBody();
		return messages;
	}

	@Override
	public List<Message> findBySenderIdAndHoroBetween(String sid, LocalDateTime horoStart, LocalDateTime horoEnd) {
		String url = msgPath + "/sid:" + sid + "/gte:" + horoStart.toString() + "/lte:" + horoEnd.toString();
		List<Message> messages = new ArrayList<>();
		MultiValueMap<String, String> headers = new LinkedMultiValueMap<>();
		headers.add("Authorization", DadakarFrontAdminApplication.getToken());
		HttpEntity<List<Message>> msgRequest = new HttpEntity<List<Message>>(messages, headers);
		ResponseEntity<List<Message>> msgResponse = restTemplate.exchange(url, HttpMethod.GET, msgRequest, new ParameterizedTypeReference<List<Message>>() {});
		messages = msgResponse.getBody();
		return messages;
	}

	@Override
	public List<Message> findByReceiverId(String rid) {
		String url = msgPath + "/rid:" + rid;
		List<Message> messages = new ArrayList<>();
		MultiValueMap<String, String> headers = new LinkedMultiValueMap<>();
		headers.add("Authorization", DadakarFrontAdminApplication.getToken());
		HttpEntity<List<Message>> msgRequest = new HttpEntity<List<Message>>(messages, headers);
		ResponseEntity<List<Message>> msgResponse = restTemplate.exchange(url, HttpMethod.GET, msgRequest, new ParameterizedTypeReference<List<Message>>() {});
		messages = msgResponse.getBody();
		return messages;
	}

	@Override
	public List<Message> findByReceiverIdAndHoroLessThanEqual(String rid, LocalDateTime horoEnd) {
		String url = msgPath + "/rid:" + rid + "/lte:" + horoEnd.toString();
		List<Message> messages = new ArrayList<>();
		MultiValueMap<String, String> headers = new LinkedMultiValueMap<>();
		headers.add("Authorization", DadakarFrontAdminApplication.getToken());
		HttpEntity<List<Message>> msgRequest = new HttpEntity<List<Message>>(messages, headers);
		ResponseEntity<List<Message>> msgResponse = restTemplate.exchange(url, HttpMethod.GET, msgRequest, new ParameterizedTypeReference<List<Message>>() {});
		messages = msgResponse.getBody();
		return messages;
	}

	@Override
	public List<Message> findByReceiverIdAndHoroGreaterThanEqual(String rid, LocalDateTime horoStart) {
		String url = msgPath + "/rid:" + rid + "/gte:" + horoStart.toString();
		List<Message> messages = new ArrayList<>();
		MultiValueMap<String, String> headers = new LinkedMultiValueMap<>();
		headers.add("Authorization", DadakarFrontAdminApplication.getToken());
		HttpEntity<List<Message>> msgRequest = new HttpEntity<List<Message>>(messages, headers);
		ResponseEntity<List<Message>> msgResponse = restTemplate.exchange(url, HttpMethod.GET, msgRequest, new ParameterizedTypeReference<List<Message>>() {});
		messages = msgResponse.getBody();
		return messages;
	}

	@Override
	public List<Message> findByReceiverIdAndHoroBetween(String rid, LocalDateTime horoStart, LocalDateTime horoEnd) {
		String url = msgPath + "/rid:" + rid + "/gte:" + horoStart.toString() + "/lte:" + horoEnd.toString();
		List<Message> messages = new ArrayList<>();
		MultiValueMap<String, String> headers = new LinkedMultiValueMap<>();
		headers.add("Authorization", DadakarFrontAdminApplication.getToken());
		HttpEntity<List<Message>> msgRequest = new HttpEntity<List<Message>>(messages, headers);
		ResponseEntity<List<Message>> msgResponse = restTemplate.exchange(url, HttpMethod.GET, msgRequest, new ParameterizedTypeReference<List<Message>>() {});
		messages = msgResponse.getBody();
		return messages;
	}

	@Override
	public List<Message> findBySenderIdAndReceiverId(String sid, String rid) {
		String url = msgPath + "/sid:" + sid + "/rid:" + rid;
		List<Message> messages = new ArrayList<>();
		MultiValueMap<String, String> headers = new LinkedMultiValueMap<>();
		headers.add("Authorization", DadakarFrontAdminApplication.getToken());
		HttpEntity<List<Message>> msgRequest = new HttpEntity<List<Message>>(messages, headers);
		ResponseEntity<List<Message>> msgResponse = restTemplate.exchange(url, HttpMethod.GET, msgRequest, new ParameterizedTypeReference<List<Message>>() {});
		messages = msgResponse.getBody();
		return messages;
	}

	@Override
	public List<Message> findBySenderIdAndReceiverIdAndHoroLessThanEqual(String sid, String rid, LocalDateTime horoEnd) {
		String url = msgPath + "/sid:" + sid + "/rid:" + rid + "/lte:" + horoEnd.toString();
		List<Message> messages = new ArrayList<>();
		MultiValueMap<String, String> headers = new LinkedMultiValueMap<>();
		headers.add("Authorization", DadakarFrontAdminApplication.getToken());
		HttpEntity<List<Message>> msgRequest = new HttpEntity<List<Message>>(messages, headers);
		ResponseEntity<List<Message>> msgResponse = restTemplate.exchange(url, HttpMethod.GET, msgRequest, new ParameterizedTypeReference<List<Message>>() {});
		messages = msgResponse.getBody();
		return messages;
	}

	@Override
	public List<Message> findBySenderIdAndReceiverIdAndHoroGreaterThanEqual(String sid, String rid, LocalDateTime horoStart) {
		String url = msgPath + "/sid:" + sid + "/rid:" + rid + "/gte:" + horoStart.toString();
		List<Message> messages = new ArrayList<>();
		MultiValueMap<String, String> headers = new LinkedMultiValueMap<>();
		headers.add("Authorization", DadakarFrontAdminApplication.getToken());
		HttpEntity<List<Message>> msgRequest = new HttpEntity<List<Message>>(messages, headers);
		ResponseEntity<List<Message>> msgResponse = restTemplate.exchange(url, HttpMethod.GET, msgRequest, new ParameterizedTypeReference<List<Message>>() {});
		messages = msgResponse.getBody();
		return messages;
	}

	@Override
	public List<Message> findBySenderIdAndReceiverIdAndHoroBetween(String sid, String rid, LocalDateTime horoStart, LocalDateTime horoEnd) {
		String url = msgPath + "/sid:" + sid + "/rid:" + rid + "/gte:" + horoStart.toString() + "/lte:" + horoEnd.toString();
		List<Message> messages = new ArrayList<>();
		MultiValueMap<String, String> headers = new LinkedMultiValueMap<>();
		headers.add("Authorization", DadakarFrontAdminApplication.getToken());
		HttpEntity<List<Message>> msgRequest = new HttpEntity<List<Message>>(messages, headers);
		ResponseEntity<List<Message>> msgResponse = restTemplate.exchange(url, HttpMethod.GET, msgRequest, new ParameterizedTypeReference<List<Message>>() {});
		messages = msgResponse.getBody();
		return messages;
	}

}

File: dadakar-front-client/src/services/message.service.ts
import { Injectable } from "@angular/core";
import { Headers, Http, Response } from "@angular/http";

import { AuthProvider } from '../providers/auth';
import { Message } from "../models/message.model";

import config from "../assets/config/config";

import 'rxjs/add/operator/map';

@Injectable()
export class MessageService {

    url: string = config.backMsgServerAddress + '/dadakar/msg/';
    private header: Headers;

    constructor(private authProvider: AuthProvider, private http:Http) {
        this.authProvider.authUser.subscribe(jwt => {
            if(jwt) {
                this.header = new Headers({
                    'Authorization': jwt.token
                });
            }
        });
    }

    add(message: Message) {
        return this.http.post(this.url + 'save', message, {headers : this.header}).map((res:Response) => res.json());
    }

    update(message: Message) {
        return this.http.put(this.url + 'update', message, {headers : this.header}).map((res:Response) => res.json());
    }

    del(messageId: string) {
        this.http.delete(this.url + 'del/' + messageId, {headers : this.header});
    }

    findAll() {
        return this.http.get(this.url, {headers : this.header}).map((res: Response) => res.json());
    }

    findById(messageId: string) {
        return this.http.get(this.url + messageId, {headers : this.header}).map((res: Response) => res.json());
    }

    findByHoroLess(end: string) {
        return this.http.get(this.url + "lte:" + end, {headers : this.header}).map((res: Response) => res.json());
    }

    findByHoroGreater(start: string) {
        return this.http.get(this.url + "gte:" + start, {headers : this.header}).map((res: Response) => res.json());
    }

    findByHoroBetween(start: string, end: string) {
        return this.http.get(this.url + "gte:" + start + "/lte:" + end, {headers : this.header}).map((res: Response) => res.json());
    }

    findBySenderId(sid: string) {
        return this.http.get(this.url + "sid:" + sid, {headers : this.header}).map((res: Response) => res.json());
    }

    findBySenderIdAndHoroLess(sid: string, end: string) {
        return this.http.get(this.url + "sid:" + sid + "/lte:" + end, {headers : this.header}).map((res: Response) => res.json());
    }

    findBySenderIdAndHoroGreater(sid: string, start: string) {
        return this.http.get(this.url + "sid:" + sid + "/gte:" + start, {headers : this.header}).map((res: Response) => res.json());
    }

    findBySenderIdAndHoroBetween(sid: string, start: string, end: string) {
        return this.http.get(this.url + "sid:" + sid + "/gte:" + start + "/lte:" + end, {headers : this.header}).map((res: Response) => res.json());
    }

    findByReceiverId(rid: string) {
        return this.http.get(this.url + "rid:" + rid, {headers : this.header}).map((res: Response) => res.json());
    }

    findByReceiverIdAndHoroLess(rid: string, end: string) {
        return this.http.get(this.url + "rid:" + rid + "/lte:" + end, {headers : this.header}).map((res: Response) => res.json());
    }

    findByReceiverIdAndHoroGreater(rid: string, start: string) {
        return this.http.get(this.url + "rid:" + rid + "/gte:" + start, {headers : this.header}).map((res: Response) => res.json());
    }

    findByReceiverIdAndHoroBetween(rid: string, start: string, end: string) {
        return this.http.get(this.url + "rid:" + rid + "/gte:" + start + "/lte:" + end, {headers : this.header}).map((res: Response) => res.json());
    }

    findBySenderIdAndReceiverId(sid: string, rid: string) {
        return this.http.get(this.url + "sid:" + sid + "/rid:" + rid, {headers : this.header}).map((res: Response) => res.json());
    }

    findBySenderIdAndReceiverIdAndHoroLess(sid: string, rid: string, end: string) {
        return this.http.get(this.url + "sid:" + sid + "/rid:" + rid + "/lte:" + end, {headers : this.header}).map((res: Response) => res.json());
    }

    findBySenderIdAndReceiverIdAndHoroGreater(sid: string, rid: string, start: string) {
        return this.http.get(this.url + "sid:" + sid + "/rid:" + rid + "/gte:" + start, {headers : this.header}).map((res: Response) => res.json());
    }

    findBySenderIdAndReceiverIdAndHoroBetween(sid: string, rid: string, start: string, end: string) {
        return this.http.get(this.url + "sid:" + sid + "/rid:" + rid + "/gte:" + start + "/lte:" + end, {headers : this.header}).map((res: Response) => res.json());
    }

}
